Crash when scraping managed Google Cloud SQL Postgres instance due to `google_dataplex.max_messages` setting

**What did you do?**

Our Cloud SQL Postgres instance updated overnight, and it now has a new setting in the `pg_settings` table.

**What did you expect to see?**

No impact.

**What did you see instead? Under which circumstances?**

The exporter now crashes with the following panic message.

```
panic: Error converting setting "google_dataplex.max_messages" value "-1:100" to float: strconv.ParseFloat: parsing "-1:100": invalid syntax

goroutine 57 [running]:
main.(*pgSetting).metric(0xc0003802d0, 0xc0000c4300)
	/app/cmd/postgres_exporter/pg_setting.go:86 +0x346
main.querySettings(0xc000085810, 0xc0000e4000)
	/app/cmd/postgres_exporter/pg_setting.go:55 +0x1d8
main.(*Server).Scrape(0xc0000e4000, 0xc000085810, 0x0?)
	/app/cmd/postgres_exporter/server.go:120 +0xc5
main.(*Exporter).scrapeDSN(0xc00015a180, 0xc000085810, {0xc000100700, 0x6a})
	/app/cmd/postgres_exporter/datasource.go:115 +0x12d
main.(*Exporter).scrape(0xc00015a180, 0xc000085810)
	/app/cmd/postgres_exporter/postgres_exporter.go:681 +0x16b
main.(*Exporter).Collect(0xc00015a180, 0xc000085810)
	/app/cmd/postgres_exporter/postgres_exporter.go:570 +0x25
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
	/go/pkg/mod/github.com/prometheus/client_golang@v1.23.1/prometheus/registry.go:458 +0xe5
created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather in goroutine 53
	/go/pkg/mod/github.com/prometheus/client_golang@v1.23.1/prometheus/registry.go:548 +0xbf1
```

**Environment**

* System information:

We are using the Docker container image.

* postgres_exporter version:

`quay.io/prometheuscommunity/postgres-exporter:v0.18.1`

* postgres_exporter flags:

```
--web.listen-address=:9187
--log.format=json
```

* PostgreSQL version:

15.15

* Logs:
```
{"time":"2026-01-16T16:32:52.926069037Z","level":"WARN","source":"main.go:85","msg":"Error loading config","err":"error opening config file \"postgres_exporter.yml\": open postgres_exporter.yml: no such file or directory"}
{"time":"2026-01-16T16:32:52.926220207Z","level":"INFO","source":"main.go:95","msg":"Excluded databases","databases":"[]"}
{"time":"2026-01-16T16:32:52.928070872Z","level":"INFO","source":"tls_config.go:346","msg":"Listening on","address":"[::]:9187"}
{"time":"2026-01-16T16:32:52.928145615Z","level":"INFO","source":"tls_config.go:349","msg":"TLS is disabled.","http2":false,"address":"[::]:9187"}
{"time":"2026-01-16T16:33:37.336956402Z","level":"INFO","source":"server.go:73","msg":"Established new database connection","fingerprint":"127.0.0.1:5432"}
{"time":"2026-01-16T16:33:37.622250713Z","level":"INFO","source":"postgres_exporter.go:615","msg":"Semantic version changed","server":{},"from":"0.0.0","to":"15.15.0"}
panic: Error converting setting "google_dataplex.max_messages" value "-1:100" to float: strconv.ParseFloat: parsing "-1:100": invalid syntax

goroutine 57 [running]:
main.(*pgSetting).metric(0xc0003802d0, 0xc0000c4300)
	/app/cmd/postgres_exporter/pg_setting.go:86 +0x346
main.querySettings(0xc000085810, 0xc0000e4000)
	/app/cmd/postgres_exporter/pg_setting.go:55 +0x1d8
main.(*Server).Scrape(0xc0000e4000, 0xc000085810, 0x0?)
	/app/cmd/postgres_exporter/server.go:120 +0xc5
main.(*Exporter).scrapeDSN(0xc00015a180, 0xc000085810, {0xc000100700, 0x6a})
	/app/cmd/postgres_exporter/datasource.go:115 +0x12d
main.(*Exporter).scrape(0xc00015a180, 0xc000085810)
	/app/cmd/postgres_exporter/postgres_exporter.go:681 +0x16b
main.(*Exporter).Collect(0xc00015a180, 0xc000085810)
	/app/cmd/postgres_exporter/postgres_exporter.go:570 +0x25
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
	/go/pkg/mod/github.com/prometheus/client_golang@v1.23.1/prometheus/registry.go:458 +0xe5
created by github.com/prometheus/client_golang/prometheus.(*Registry).Gather in goroutine 53
	/go/pkg/mod/github.com/prometheus/client_golang@v1.23.1/prometheus/registry.go:548 +0xbf1

```

**Other information**

The new pg_settings row has the following values.

```
             name             | setting | unit |      category      |                    short_desc                    | extra_desc |  context   | vartype | source  | min_val | max_val | enumvals | boot_val | reset_val | sourcefile | sourceline | pending_restart 
------------------------------+---------+------+--------------------+--------------------------------------------------+------------+------------+---------+---------+---------+---------+----------+----------+-----------+------------+------------+-----------------
 google_dataplex.max_messages | -1:100  |      | Customized Options | Maximum number of pending notification messages. |            | postmaster | integer | default | -1      | 65536   |          | -1       | -1        |            |            | f
```

I noticed there is already an exception in the relevant query for an Azure database setting with `vartype = 'bool'` that has unexpected settings values. See #523 and #587. We could either fix this by adding another exception, or return a NaN in this situation. The existing code errs on the side of failing on unexpected values, in order to avoid missing metrics.

--- a/exporter/pg_setting.go
+++ b/exporter/pg_setting.go
@@ -51,8 +51,15 @@ func querySettings(ch chan<- prometheus.Metric, server *Server) error {
 		if err != nil {
 			return fmt.Errorf("Error retrieving rows on %q: %s %v", server, namespace, err)
 		}
-
-		ch <- s.metric(server.labels)
+		metric, err := s.metric(server.labels)
+		if err != nil {
+			// Log the error and continue
+			// This could be due to a bad value in the setting
+			// but we should not fail the entire scrape or panic
+			server.logger.Warn("Error normalising unit for setting", "setting", s.name, "value", s.setting, "unit", s.unit, "error", err)
+			continue
+		}
+		ch <- metric
 	}
 
 	return nil
@@ -64,7 +71,7 @@ type pgSetting struct {
 	name, setting, unit, shortDesc, vartype string
 }
 
-func (s *pgSetting) metric(labels prometheus.Labels) prometheus.Metric {
+func (s *pgSetting) metric(labels prometheus.Labels) (prometheus.Metric, error) {
 	var (
 		err       error
 		name      = strings.ReplaceAll(strings.ReplaceAll(s.name, ".", "_"), "-", "_")
@@ -81,9 +88,7 @@ func (s *pgSetting) metric(labels prometheus.Labels) prometheus.Metric {
 		}
 	case "integer", "real":
 		if val, unit, err = s.normaliseUnit(); err != nil {
-			// Panic, since we should recognise all units
-			// and don't want to silently exlude metrics
-			panic(err)
+			return nil, err
 		}
 
 		if len(unit) > 0 {
@@ -92,11 +97,11 @@ func (s *pgSetting) metric(labels prometheus.Labels) prometheus.Metric {
 		}
 	default:
 		// Panic because we got a type we didn't ask for
-		panic(fmt.Sprintf("Unsupported vartype %q", s.vartype))
+		return nil, fmt.Errorf("Unsupported vartype %q", s.vartype)
 	}
 
 	desc := newDesc(subsystem, name, shortDesc, labels)
-	return prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, val)
+	return prometheus.MustNewConstMetric(desc, prometheus.GaugeValue, val), nil
 }
 
 // Removes units from any of the setting values.

--- a/exporter/pg_setting.go
+++ b/exporter/pg_setting.go
@@ -96,7 +96,6 @@ func (s *pgSetting) metric(labels prometheus.Labels) (prometheus.Metric, error) 
 			shortDesc = fmt.Sprintf("%s [Units converted to %s.]", shortDesc, unit)
 		}
 	default:
-		// Panic because we got a type we didn't ask for
 		return nil, fmt.Errorf("pgsetting: unsupported vartype %q", s.vartype)
 	}
 

--- a/exporter/pg_setting.go
+++ b/exporter/pg_setting.go
@@ -97,7 +97,7 @@ func (s *pgSetting) metric(labels prometheus.Labels) (prometheus.Metric, error) 
 		}
 	default:
 		// Panic because we got a type we didn't ask for
-		return nil, fmt.Errorf("Unsupported vartype %q", s.vartype)
+		return nil, fmt.Errorf("pgsetting: unsupported vartype %q", s.vartype)
 	}
 
 	desc := newDesc(subsystem, name, shortDesc, labels)

--- a/exporter/pg_setting_test.go
+++ b/exporter/pg_setting_test.go
@@ -237,17 +237,15 @@ func (s *PgSettingSuite) TestNormaliseUnit(c *C) {
 }
 
 func (s *PgSettingSuite) TestMetric(c *C) {
-	defer func() {
-		if r := recover(); r != nil {
-			if r.(error).Error() != `unknown unit for runtime variable: "nonexistent"` {
-				panic(r)
-			}
-		}
-	}()
-
 	for _, f := range fixtures {
+		if f.n.err != "" {
+			continue
+		}
 		d := &dto.Metric{}
-		m := f.p.metric(prometheus.Labels{})
+		m, err := f.p.metric(prometheus.Labels{})
+		if err != nil {
+			c.Fatalf("Error creating metric: %v", err)
+		}
 		m.Write(d) // nolint: errcheck
 
 		c.Check(m.Desc().String(), Equals, f.d)

